swarm/src/behaviour.rs

```diff
@@ -121,6 +121,14 @@ pub trait NetworkBehaviour {
     fn inject_new_external_addr(&mut self, _addr: &Multiaddr) {
     }
 
+    /// A listener experienced an error.
+    fn inject_listener_error(&mut self, _id: ListenerId, _err: &dyn std::error::Error) {
```

Comment:
I'm wondering how useful a `&dyn Error` is, i.e. isn't the only thing that can be done with it to print it via `Debug` or `Display`? All methods on `std::error::Error` that allow inspecting the error (`downcast` / `is`) seem to require the `'static` lifetime. I would expect that a main reason to pass an error to a callback is to allow the callback to potentially inspect and handle it according to the type of the error.

Reply:
I.e. wouldn't a custom enum `ListenerError` be desirable?

Reply:
The `'static` bound should be added I think. As for 

> I.e. wouldn't a custom enum `ListenerError` be desirable?

What should the `ListenerError` enum contain?

Reply:
> What should the `ListenerError` enum contain?

To start with it would probably just be
```rust
#[non_exhaustive]
pub enum ListenError {
    /// A listener failed due to an error encountered by the underlying transport.
    ///
    /// A transport error is often transient in nature, e.g. an I/O error.
    TransportError(Box<dyn Error + 'static>),
}
```
with the possibility to add `libp2p-swarm`-specific error cases later. I'm just mentioning it, because I've experimented with something similar for `inject_addr_reach_failure`, which currently also just takes a `&dyn Error`, [here](https://github.com/libp2p/rust-libp2p/commit/bcebde7b28c5a7b941ed7f86f8d9893d869d5092#diff-bd8e7718bb252fedaadead082bed8489R134). In that case, there are already `libp2p-swarm`-specific error cases other than a transport error. I see a few, admittedly small, advantages of an indirection through an explicit enum:

  1. It "annotates" the opaque error as a transport error, thereby documenting through the types that it makes sense to try to downcast / inspect the error as a e.g. TCP transport error.
  2. A little future-proving of the API with added type information, i.e. if there are other error cases in the future, they can be added through an explicit new constructor, instead of having to document that the `dyn Error` may be either a transport error or one of possibly a number of `libp2p-swarm`-specific errors.
  3. If the `DialError` idea linked to above is added later as well, there would be some symmetry.

This is just a suggestion that could also be investigated further later and need not delay merging this PR!

Reply:
I would like to keep it as a trait object for the time being because it is rare in my experience that programmes base decisions on specific error cases instead of just failing if an error of any kind is encountered. But it could of course be the topic of another PR I guess. What is your take on this, @tomaka?

Reply:
I don't really have a strong opinion either. I would keep the `&dyn Error` because adding yet another `ListenerError` item adds confusion.